horovod/common/controller.cc

```diff
@@ -658,8 +668,8 @@ ResponseList Controller::FuseResponses(std::deque<Response>& responses,
 
         std::vector<TensorTableEntry> entries_for_join;
         if (joined) {
-          tensor_queue_.GetTensorEntriesFromResponse(response, entries_for_join,
-                                                     joined);
+          tensor_queue_.GetTensorEntriesFromResponse(new_response,
+                                                     entries_for_join, joined);
```

Comment:
Thanks for the quick PR to fix fusion!

@tgaddair and I were discussing this and we think the fusion code here can be simplified a bit. Would it be possible to get both the `dtype` and `new_tensor_size` directly from the `new_response`, instead of generating `new_entry`? This would remove special treatment for join in `FuseResponses` and the code below to determine if fusion is possible would look something like:
```
        int64_t new_tensor_size = new_response.tensor_sizes()[0];
        if (response.response_type() == new_response.response_type() &&
            response.devices() == new_response.devices() &&
            response.tensor_type() == new_response.tensor_type() &&
            tensor_size + new_tensor_size <= TensorFusionThresholdBytes())
```
This would require making sure the type and tensor size fields are set in the response are set unconditionally here (and also for ADASUM): https://github.com/horovod/horovod/blob/master/horovod/common/controller.cc#L585
but we think that is alright to do.

With this, you can also remove the method `tensor_queue_.GetTensorSizeAndType`, since I don't think it is used anywhere else.

Reply:
Sent an update with these changes.

Reply:
I've realized that there are couple of other things I can remove, and `FuseResponses()` doesn't even need `joined` parameter. Doing this now...

Reply:
Updated.